policy-controller/k8s/index/src/index.rs

```diff
@@ -0,0 +1,738 @@
+                // state actually changed. The vast majority of resets will see
+                // no actual data change.
+                self.namespaces
+                    .get_or_default_with_reindex(namespace, |ns| {
+                        let mut changed = !removed.is_empty();
+                        for name in removed.into_iter() {
+                            ns.policy.servers.remove(&name);
+                        }
+                        for (name, server) in added.into_iter() {
+                            changed = ns.policy.update_server(name, server) || changed;
+                        }
+                        changed
+                    });
+            }
+        }
+    }
+}
+
+impl kubert::index::IndexNamespacedResource<k8s::policy::ServerAuthorization> for Index {
+    #[instrument(skip_all, name = "saz.apply", fields(
+        ns = %saz.namespace().unwrap(),
+        saz = %saz.name(),
+    ))]
+    fn apply(&mut self, saz: k8s::policy::ServerAuthorization) {
+        let namespace = saz.namespace().unwrap();
+        let name = saz.name();
+        match server_authorization::ServerAuthz::from_resource(saz, &self.cluster_info) {
+            Ok(meta) => self
+                .namespaces
+                .get_or_default_with_reindex(namespace, move |ns| {
+                    ns.policy.update_server_authz(name, meta)
+                }),
+            Err(error) => tracing::error!(%error, "Illegal server authorization update"),
+        }
+    }
+
+    #[instrument(skip(self), name = "saz.delete")]
+    fn delete(&mut self, ns: String, saz: String) {
+        self.namespaces.get_with_reindex(ns, |ns| {
+            ns.policy.server_authorizations.remove(&saz).is_some()
+        })
+    }
+
+    #[instrument(skip_all, name = "saz.reset")]
+    fn reset(
+        &mut self,
+        sazs: Vec<k8s::policy::ServerAuthorization>,
+        deleted: HashMap<String, HashSet<String>>,
+    ) {
+        #[derive(Default)]
+        struct Ns {
+            added: Vec<(String, server_authorization::ServerAuthz)>,
+            removed: HashSet<String>,
+        }
+
+        // Aggregate all of the updates by namespace so that we only reindex
+        // once per namespace.
+        let mut updates_by_ns = HashMap::<String, Ns>::default();
+        for saz in sazs.into_iter() {
+            let namespace = saz
+                .namespace()
+                .expect("serverauthorization must be namespaced");
+            let name = saz.name();
+            match server_authorization::ServerAuthz::from_resource(saz, &self.cluster_info) {
+                Ok(saz) => updates_by_ns
+                    .entry(namespace)
+                    .or_default()
+                    .added
+                    .push((name, saz)),
+                Err(error) => {
+                    tracing::error!(%namespace, saz = %name, %error, "Illegal server authorization update")
```

Comment:
could remove the ns/serverauthorization fields from this if they were included in the span and the span was at the ERROR level, but TIOLI.

Reply:
I don't think that's any better here, since that span would have only that single event in it.